sdk/inc/azure/iot/az_iot_pnp_client.h

```diff
@@ -0,0 +1,706 @@
+    az_iot_pnp_client const* client,
+    uint64_t token_expiration_epoch_time,
+    az_span base64_hmac_sha256_signature,
+    az_span key_name,
+    char* mqtt_password,
+    size_t mqtt_password_size,
+    size_t* out_mqtt_password_length);
+
+/*
+ *
+ * PnP Telemetry APIs
+ *
+ */
+
+/**
+ * @brief Gets the MQTT topic that is used for device to cloud telemetry messages.
+ * @note This topic can also be used to set the MQTT Will message in the Connect message.
+ *
+ * @param[in] client The #az_iot_pnp_client to use for this call.
+ * @param[in] component_name An #az_span specifying the component name to publish telemetry on.
+ * @param[in] properties Properties to attach to append to the topic.
+ * @param[out] mqtt_topic A buffer with sufficient capacity to hold the MQTT topic. If successful,
+ * contains a null-terminated string with the topic that needs to be passed to the MQTT client.
+ * @param[in] mqtt_topic_size The size, in bytes, of \p mqtt_topic.
+ * @param[out] out_mqtt_topic_length __[nullable]__ Contains the string length, in bytes, of \p
+ * mqtt_topic. Can be `NULL`.
+ *
+ * @pre \p client must not be `NULL`.
+ * @pre \p mqtt_topic must not be `NULL`.
+ * @pre \p mqtt_topic_size must be greater than 0.
+ *
+ * @return An #az_result value indicating the result of the operation.
+ * @retval #AZ_OK The topic was retrieved successfully.
+ */
+AZ_NODISCARD az_result az_iot_pnp_client_telemetry_get_publish_topic(
+    az_iot_pnp_client const* client,
+    az_span component_name,
+    az_iot_message_properties* properties,
+    char* mqtt_topic,
+    size_t mqtt_topic_size,
+    size_t* out_mqtt_topic_length);
+
+/*
+ *
+ * PnP Command APIs
+ *
+ */
+
+/**
+ * @brief The MQTT topic filter to subscribe to command requests.
+ * @note Commands MQTT Publish messages will have QoS At most once (0).
+ */
+#define AZ_IOT_PNP_CLIENT_COMMANDS_SUBSCRIBE_TOPIC "$iothub/methods/POST/#"
+
+/**
+ * @brief A command request received from IoT Hub.
+ *
+ */
+typedef struct
+{
+  /**
+   * The request ID.
+   * @note The application must match the command request and command response.
+   */
+  az_span request_id;
+
+  /**
+   * The name of the component which the command was invoked for.
+   * @note Can be `AZ_SPAN_EMPTY` if for the root component
+   */
+  az_span component_name;
+
+  /**
+   * The command name.
+   */
+  az_span command_name;
+} az_iot_pnp_client_command_request;
+
+/**
+ * @brief Attempts to parse a received message's topic.
+ *
+ * @param[in] client The #az_iot_pnp_client to use for this call.
+ * @param[in] received_topic An #az_span containing the received topic.
+ * @param[out] out_request If the message is a command request, this will contain the
+ * #az_iot_pnp_client_command_request.
+ *
+ * @pre \p client must not be `NULL`.
+ * @pre \p received_topic must be a valid, non-empty #az_span.
+ * @pre \p out_request must not be `NULL`. It must point to an #az_iot_pnp_client_command_request
```

Comment:
Same here/elsewhere. These comments aren't necessary in a type-safe language. It would only be useful if we were using things like `void*`.
```suggestion
 * @pre \p out_request must not be `NULL`.
```